okhttp/src/main/java/com/lightstep/tracer/shared/HttpCollectorClient.java

```diff
@@ -2,18 +2,99 @@
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
 
 class HttpCollectorClient extends CollectorClient {
-  HttpCollectorClient() {}
+    private static final MediaType protoMediaType = MediaType.parse("application/octet-stream");
+
+    private final AtomicReference<OkHttpClient> client;
+    private final AbstractTracer tracer;
+    private final URL collectorURL;
+    private final long deadlineMillis;
+
+    HttpCollectorClient(
+            AbstractTracer tracer,
+            URL collectorURL,
+            long deadlineMillis
+    ) {
+        this.client = new AtomicReference<>(start(deadlineMillis));
+        this.tracer = tracer;
+        this.collectorURL = collectorURL;
+        this.deadlineMillis = deadlineMillis;
+    }
+
+    @Override
+    ReportResponse report(ReportRequest request) {
+        try {
+            return fromResponse(this.client().newCall(toRequest(request)).execute());
+        } catch (Exception e) {
+            tracer.error("Exception sending report to collector: ", e);
+            return null;
+        }
+    }
+
+    @Override
+    void reconnect() {
+        shutdown(this.client.getAndSet(start(this.deadlineMillis)));
+    }
+
+    @Override
+    void shutdown() {
+        shutdown(this.client());
+    }
+
+    private OkHttpClient client() {
+        return this.client.get();
+    }
+
+    private Request toRequest(ReportRequest request) {
+        return new Request.Builder()
+                .url(this.collectorURL)
+                .post(RequestBody.create(protoMediaType, request.toByteArray()))
+                .build();
+    }
+
+    private ReportResponse fromResponse(Response response) throws IOException {
+        try {
+            if (!response.isSuccessful()) {
+                this.tracer.error(String.format(
+                        "Collector returned non-successful http code %d",
+                        response.code()
+                ));
+                return null;
+            }
+
+            return fromResponseBody(response.body());
+        } finally {
+            response.close();
+        }
+    }
 
-  @Override
-  ReportResponse report(ReportRequest request) {
-    return ReportResponse.getDefaultInstance();
-  }
+    private ReportResponse fromResponseBody(ResponseBody body) throws IOException {
+        if (body == null) {
+            this.tracer.error("Collector returned an empty body");
+            return null;
+        }
+        return ReportResponse.parseFrom(body.byteStream());
+    }
 
-  @Override
-  void reconnect() {}
+    private static void shutdown(OkHttpClient client) {
+        client.dispatcher().executorService().shutdown();
+    }
 
-  @Override
-  void shutdown() {}
+    private static OkHttpClient start(long deadlineMillis) {
+        return new OkHttpClient.Builder()
+                .connectTimeout(deadlineMillis, TimeUnit.MILLISECONDS)
```

Comment:
I'm not 100% sure this is the right timeout we want.

Reply:
seems right